File: userspace/wait_diff/openat2_resolve.c
/* openat2(2) RESOLVE_* scoping — does the sandbox hold on the CREATE path?
 *
 * `fs/open.c` build_open_flags folds every RESOLVE_* bit into ONE
 * `op->lookup_flags`, and `fs/namei.c` path_openat runs the LOOKUP_PARENT
 * walk that O_CREAT needs with the SAME `nd->flags`. There is no create-path
 * exception, so a scoping bit that stops constraining resolution the moment a
 * create is involved is a sandbox escape.
 *
 * Every case is rooted in a private mkdtemp() and reports only the OUTCOME
 * (named errno + whether the escape target exists), never a path, so the host
 * oracle and the guest print identical records regardless of uid or /tmp
 * contents.
 *
 * Layout built per case:
 *   <base>/box          the dirfd — the scope
 *   <base>/box/sub      an ordinary child, inside the scope
 *   <base>/outside      the escape target, one level ABOVE the scope
 */
#include "probe.h"

#ifndef RESOLVE_NO_XDEV
#define RESOLVE_NO_XDEV       0x01
#endif
#ifndef RESOLVE_NO_MAGICLINKS
#define RESOLVE_NO_MAGICLINKS 0x02
#endif
#ifndef RESOLVE_NO_SYMLINKS
#define RESOLVE_NO_SYMLINKS   0x04
#endif
#ifndef RESOLVE_BENEATH
#define RESOLVE_BENEATH       0x08
#endif
#ifndef RESOLVE_IN_ROOT
#define RESOLVE_IN_ROOT       0x10
#endif

/* glibc has no openat2 wrapper on either arch; the raw slot is the ABI under
 * test anyway (the wrapper would only re-pack the same struct). */
/* mkdtemp template is 17 bytes; a small fixed base keeps every derived
 * snprintf provably short of PATH_MAX (the probe builds with -Werror). */
#define O2_BASE_MAX 32

struct o2_how { uint64_t flags; uint64_t mode; uint64_t resolve; };

static int o2(int dfd, const char *path, uint64_t flags, uint64_t mode, uint64_t resolve) {
    struct o2_how how;
    memset(&how, 0, sizeof how);
    how.flags = flags;
    how.mode = mode;
    how.resolve = resolve;
    return (int)syscall(SYS_openat2, dfd, path, &how, sizeof how);
}

/* Build <base>/{box,box/sub,outside} and return an O_PATH-free dirfd on box.
 * `base` must be PATH_MAX-sized; it receives the mkdtemp() result. */
static int make_scope(char *base, size_t baselen) {
    char p[O2_BASE_MAX + 32];
    snprintf(base, baselen, "/tmp/o2res.XXXXXX");
    if (!mkdtemp(base)) return -1;
    snprintf(p, sizeof p, "%s/box", base);     if (mkdir(p, 0700) < 0) return -1;
    snprintf(p, sizeof p, "%s/box/sub", base); if (mkdir(p, 0700) < 0) return -1;
    snprintf(p, sizeof p, "%s/outside", base); if (mkdir(p, 0700) < 0) return -1;
    snprintf(p, sizeof p, "%s/box", base);
    return open(p, O_RDONLY | O_DIRECTORY);
}

static int exists(const char *base, const char *rel) {
    char p[O2_BASE_MAX + 32];
    struct stat st;
    snprintf(p, sizeof p, "%s/%s", base, rel);
    return stat(p, &st) == 0 ? 1 : 0;
}

static void cleanup(const char *base) {
    char p[O2_BASE_MAX + 32];
    snprintf(p, sizeof p, "%s/outside/esc", base);  unlink(p);
    snprintf(p, sizeof p, "%s/box/esc", base);      unlink(p);
    snprintf(p, sizeof p, "%s/box/sub/esc", base);  unlink(p);
    snprintf(p, sizeof p, "%s/box/link", base);     unlink(p);
    snprintf(p, sizeof p, "%s/box/dangle", base);   unlink(p);
    snprintf(p, sizeof p, "%s/box/sub", base);      rmdir(p);
    snprintf(p, sizeof p, "%s/box", base);          rmdir(p);
    snprintf(p, sizeof p, "%s/outside", base);      rmdir(p);
    rmdir(base);
}

/* One create attempt: report the named errno and whether the file landed
 * OUTSIDE the scope. `escaped=1` on either kernel is the defect. */
static void create_case(const char *test, const char *path, uint64_t resolve) {
    char base[O2_BASE_MAX];
    int dfd = make_scope(base, sizeof base);
    int fd, err, esc;
    if (dfd < 0) { out("openat2", test, "setup=FAIL"); return; }
    errno = 0;
    fd = o2(dfd, path, O_WRONLY | O_CREAT, 0600, resolve);
    err = errno;
    if (fd >= 0) close(fd);
    esc = exists(base, "outside/esc");
    out("openat2", test, "ret=%s|errno=%s|escaped=%d|in_scope=%d",
        fd >= 0 ? "ok" : "err", fd >= 0 ? "OK" : errno_name(err), esc,
        exists(base, "box/esc") || exists(base, "box/sub/esc"));
    close(dfd);
    cleanup(base);
}

/* Same, but the path traverses a symlink planted at <base>/box/link that
 * points OUT of the scope. */
static void create_case_via_symlink(const char *test, const char *path, uint64_t resolve) {
    char base[O2_BASE_MAX], p[PATH_MAX];
    int dfd = make_scope(base, sizeof base);
    int fd, err;
    if (dfd < 0) { out("openat2", test, "setup=FAIL"); return; }
    snprintf(p, sizeof p, "%s/box/link", base);
    if (symlink("../outside", p) < 0) { out("openat2", test, "setup=FAIL"); close(dfd); cleanup(base); return; }
    errno = 0;
    fd = o2(dfd, path, O_WRONLY | O_CREAT, 0600, resolve);
    err = errno;
    if (fd >= 0) close(fd);
    out("openat2", test, "ret=%s|errno=%s|escaped=%d",
        fd >= 0 ? "ok" : "err", fd >= 0 ? "OK" : errno_name(err),
        exists(base, "outside/esc"));
    close(dfd);
    cleanup(base);
}

void probe_openat2_resolve(void) {
    /* openat2 itself must exist before any scoping claim means anything. */
    {
        char base[O2_BASE_MAX];
        int dfd = make_scope(base, sizeof base);
        int fd = dfd < 0 ? -1 : o2(dfd, "sub", O_RDONLY, 0, 0);
        int err = errno;
        out("openat2", "supported", "ret=%s|errno=%s",
            fd >= 0 ? "ok" : "err", fd >= 0 ? "OK" : errno_name(err));
        if (fd >= 0) close(fd);
        if (dfd >= 0) { close(dfd); cleanup(base); }
    }

    /* Control: an ordinary in-scope create must still work under each
     * scoping bit, so a "refuses everything" kernel cannot read as a pass. */
    create_case("in_root_create_inside",  "sub/esc", RESOLVE_IN_ROOT);
    create_case("beneath_create_inside",  "sub/esc", RESOLVE_BENEATH);
    create_case("nosym_create_inside",    "sub/esc", RESOLVE_NO_SYMLINKS);

    /* THE ESCAPE. RESOLVE_IN_ROOT *clamps* `..` at the dirfd rather than
     * erroring, so the scoped walk of `../outside/esc` ends at
     * box/outside/esc → ENOENT. A kernel that then re-resolves the PARENT
     * unscoped lands on <base>/outside and creates there: escaped=1. */
    create_case("in_root_create_dotdot", "../outside/esc", RESOLVE_IN_ROOT);

    /* Same escape reached by an ABSOLUTE pathname: under IN_ROOT the dirfd is
     * "/", so any absolute path must restart INSIDE the box. Unscoped parent
     * resolution instead honours the real root. `/tmp` exists on both kernels;
     * the box has no `tmp` child, so the correct answer is ENOENT. */
    create_case("in_root_create_absolute", "/tmp/o2res-nonexistent/esc", RESOLVE_IN_ROOT);

    /* RESOLVE_BENEATH errors instead of clamping: `..` at the scoped root is
     * EXDEV, and an absolute pathname is EXDEV, create or not. */
    create_case("beneath_create_dotdot",   "../outside/esc", RESOLVE_BENEATH);
    create_case("beneath_create_absolute", "/tmp/o2res-nonexistent/esc", RESOLVE_BENEATH);

    /* RESOLVE_NO_SYMLINKS: ANY symlink anywhere in the walk is ELOOP — the
     * parent walk of a create included. */
    create_case_via_symlink("nosym_create_via_symlink", "link/esc", RESOLVE_NO_SYMLINKS);
    /* ...and with no scoping bit the same path is the baseline that proves the
     * symlink really does lead out of the box (escaped=1 is CORRECT here). */
    create_case_via_symlink("unscoped_create_via_symlink", "link/esc", 0);

    /* RESOLVE_BENEATH must also refuse the symlink escape (its target is
     * relative and walks `..` above the scoped root → EXDEV). */
    create_case_via_symlink("beneath_create_via_symlink", "link/esc", RESOLVE_BENEATH);

    /* Unknown resolve bits and the mutually-exclusive scoping pair are EINVAL
     * (`build_open_flags`), checked before any walk happens. */
    {
        char base[O2_BASE_MAX];
        int dfd = make_scope(base, sizeof base);
        int fd, err;
        if (dfd < 0) { out("openat2", "resolve_validation", "setup=FAIL"); return; }
        errno = 0; fd = o2(dfd, "sub", O_RDONLY, 0, 0x40);
        err = errno;
        out("openat2", "unknown_resolve_bit", "ret=%s|errno=%s",
            fd >= 0 ? "ok" : "err", fd >= 0 ? "OK" : errno_name(err));
        if (fd >= 0) close(fd);
        errno = 0; fd = o2(dfd, "sub", O_RDONLY, 0, RESOLVE_BENEATH | RESOLVE_IN_ROOT);
        err = errno;
        out("openat2", "beneath_plus_in_root", "ret=%s|errno=%s",
            fd >= 0 ? "ok" : "err", fd >= 0 ? "OK" : errno_name(err));
        if (fd >= 0) close(fd);
        close(dfd);
        cleanup(base);
    }
}

File: crates/kernel/syscalls/src/openat2_resolve.rs
//! `openat2(2)` slot 437 — the `struct open_how::resolve` word: its validation
//! and its mapping onto `vfs::LookupFlags`, for BOTH walk phases.
//!
//! Lives outside the kernel-only slot file because `257_openat.rs` is
//! `#![cfg(target_os = "oxide-kernel")]`, which silently compiles any
//! `#[cfg(test)] mod tests` inside it away (CLAUDE.md phantom-test rule,
//! `docs/53`). This is the decision that determines whether an `openat2`
//! sandbox actually holds, so it is the last thing that may ship untested.
//!
//! Linux keeps every `RESOLVE_*` bit in `op->lookup_flags`
//! (`build_open_flags`) and hands that single word to
//! `do_filp_open` → `path_openat`, which uses it for the WHOLE walk —
//! the `LOOKUP_PARENT` phase that `O_CREAT` needs included
//! (Linux's `path_openat` → `link_path_walk` with `nd->flags`).
//! There is no branch on which `RESOLVE_*` bits survive a create.

use syscall::errno::Errno;

// openat2 UAPI `RESOLVE_*` bits.
pub const RESOLVE_NO_XDEV: u64 = 0x01;
pub const RESOLVE_NO_MAGICLINKS: u64 = 0x02;
pub const RESOLVE_NO_SYMLINKS: u64 = 0x04;
pub const RESOLVE_BENEATH: u64 = 0x08;
pub const RESOLVE_IN_ROOT: u64 = 0x10;
pub const RESOLVE_CACHED: u64 = 0x20;
/// `VALID_RESOLVE_FLAGS`.
pub const RESOLVE_VALID: u64 = RESOLVE_NO_XDEV | RESOLVE_NO_MAGICLINKS | RESOLVE_NO_SYMLINKS
    | RESOLVE_BENEATH | RESOLVE_IN_ROOT | RESOLVE_CACHED;

/// `build_open_flags` resolve-word admission: unknown bits are
/// `EINVAL`, and the two scoping flags are mutually exclusive.
/// # C: O(1)
pub fn validate_resolve(resolve: u64) -> Result<(), Errno> {
    if resolve & !RESOLVE_VALID != 0 { return Err(Errno::Einval); }
    if (resolve & RESOLVE_BENEATH != 0) && (resolve & RESOLVE_IN_ROOT != 0) { return Err(Errno::Einval); }
    Ok(())
}

/// `RESOLVE_*` → `LookupFlags`, mirroring the `lookup_flags |= LOOKUP_*`
/// ladder at the tail of `build_open_flags`.
/// # C: O(1)
pub fn lookup_flags_from_resolve(resolve: u64) -> vfs::LookupFlags {
    vfs::LookupFlags {
        no_xdev:       resolve & RESOLVE_NO_XDEV != 0,
        no_magiclinks: resolve & RESOLVE_NO_MAGICLINKS != 0,
        no_symlinks:   resolve & RESOLVE_NO_SYMLINKS != 0,
        beneath_exdev: resolve & RESOLVE_BENEATH != 0,
        in_root:       resolve & RESOLVE_IN_ROOT != 0,
        cached:        resolve & RESOLVE_CACHED != 0,
        ..Default::default()
    }
}

/// True when any `RESOLVE_*` modifier is set, so the open takes the flag-aware
/// resolve route that surfaces `EXDEV`/`ELOOP`/`EAGAIN` instead of collapsing
/// an escape attempt to `ENOENT`. # C: O(1)
pub fn resolve_active(x: &vfs::LookupFlags) -> bool {
    x.no_xdev || x.no_magiclinks || x.no_symlinks || x.beneath_exdev || x.in_root || x.cached
}

/// `LookupFlags` for the `LOOKUP_PARENT` phase of an `O_CREAT` open.
///
/// Every scoping bit SURVIVES: Linux never rebuilds `nd->flags` between the
/// parent walk and the final component, so `RESOLVE_BENEATH`/`IN_ROOT`/
/// `NO_SYMLINKS`/`NO_MAGICLINKS`/`NO_XDEV` constrain the directory the new
/// file is created in exactly as they constrain an ordinary open. Dropping
/// them is a sandbox escape: `openat2(dirfd, "/etc/x", O_CREAT, {RESOLVE_IN_ROOT})`
/// would create at the REAL `/etc`.
///
/// Three bits deliberately do not carry over, matching where Linux applies
/// them:
/// - `no_follow_final` (`O_NOFOLLOW`) scopes the FINAL component only; the
///   parent walk's own last component must still be followed.
/// - `directory` (`LOOKUP_DIRECTORY`) likewise constrains the final component,
///   and `O_DIRECTORY|O_CREAT` is `EINVAL` before this point anyway.
/// - `cached` (`RESOLVE_CACHED`) cannot coexist with `O_CREAT` — `EAGAIN` is
///   returned by `build_open_flags` long before any walk.
/// # C: O(1)
pub fn parent_lookup_flags(x: &vfs::LookupFlags) -> vfs::LookupFlags {
    vfs::LookupFlags {
        parent:        true,
        no_xdev:       x.no_xdev,
        no_magiclinks: x.no_magiclinks,
        no_symlinks:   x.no_symlinks,
        beneath_exdev: x.beneath_exdev,
        in_root:       x.in_root,
        ..Default::default()
    }
}

#[cfg(test)]
mod tests;
